### ISSUE TITLE
Allow customising memory limits for Cypress workers for lower memory CI instances.

### ISSUE THREAD

**Issue by docwhite** [nrwl/nx#6288]
## Description

I would like to be able to reduce the amount of memory workers take when using the plugin. At the moment, when I run `nx run myapp:e2e` it prints `Using 1 worker with 2048MB memory limit`.

```
- Generating browser application bundles (phase: setup)...
[...]
** Angular Live Development Server is listening on localhost:4200, open your browser on http://localhost:4200/ **
✔ Compiled successfully.
[...]
====================================================================================================

  (Run Starting)

  ┌────────────────────────────────────────────────────────────────────────────────────────────────┐
  │ Cypress:    7.5.0                                                                              │
  │ Browser:    Electron 89 (headless)                                                             │
  │ Specs:      3 found (modules/login/login/login-logout-redirect.spec.ts, modules/login/login/lo │
  │             gin.spec.ts, modules/login/logout/logout.spec.ts)                                  │
  └────────────────────────────────────────────────────────────────────────────────────────────────┘


────────────────────────────────────────────────────────────────────────────────────────────────────
                                                                                                    
  Running:  modules/login/login/login-logout-redirect.spec.ts                               (1 of 3)
Starting type checking service...
Using 1 worker with 2048MB memory limit
Starting type checking service...
Using 1 worker with 2048MB memory limit
```

- I would like to make each worker take 1024MB instead of 2048

## Motivation

That way we could have users run their Cypress CI processes on lower memory instances.

I am running a GitLab Runner with 1 GB of RAM configured with 1 GB of swap space (AWS T3 Micro).

## Suggested Implementation

For that to work we would need to tap into `ForkTsCheckerWebpackPlugin`'s constructor in order to inject the `memoryLimit` option with Nx.

Here's the relevant file: https://github.com/nrwl/nx/blob/12.5.7/packages/cypress/src/plugins/preprocessor.ts#L53

```javascript
      new ForkTsCheckerWebpackPlugin({
        tsconfig: config.env.tsConfig,
        useTypescriptIncrementalApi: false,
        ...suggestingDesigningSomeConfigurationObjectThatCanGetExpandedHere // <------- Customisable object!
      }),

```

So that way I could pipe the memoryLimit there as:

```javascript
suggestingDesigningSomeConfigurationObjectThatCanGetExpandedHere = { memoryLimit: 1024 }; // <----- Customisable object!
```

## Alternate Implementations

Maybe we could only target that `memoryLimit` option, since more people that are running pipelines on budget could have end to end CI without running out of memory or crashing the instance.

**Comment by skydever** [comment]
Hi there!

I am having issues with the memory consumption within cypress e2e tests as well.
After upgrading to nx 12.5.7 my CI jobs for e2e tests started failing with an error message like that:

```
<--- Last few GCs --->
[652:0x1ac600000000]  2410614 ms: Mark-sweep (reduce) 3410.7 (3702.2) -> 3335.4 (3618.8) MB, 3906.9 / 0.1 ms  (average mu = 0.763, current mu = 0.151) last resort GC in old space requested
[652:0x1ac600000000]  2415044 ms: Mark-sweep (reduce) 3335.4 (3618.8) -> 3334.9 (3586.3) MB, 4429.4 / 0.1 ms  (average mu = 0.599, current mu = 0.000) last resort GC in old space requested
<--- JS stacktrace --->
FATAL ERROR: CALL_AND_RETRY_LAST Allocation failed - JavaScript heap out of memory
```

I think being able to pass the `memoryLimit` option for cypress jobs would fix my issue too?

I am having *enough* memory on my CI environment, but I think I reach some other limits. I also tried to set `NODE_OPTIONS=--max-old-space-size=10240` (very high value), but that did not solve my problems - maybe because now I really reach my CI limits (if there are multiple processes with that 10GB limit).

I am also wondering about the `enabled` flag of `ForkTsCheckerWebpackPlugin`. Would it disable type checking at all and save huge amounts of memory (at least in my case) when running cypress tests? As far as I understand, the type checking while executing cypress tests is not very useful (or even useless?).

I have separate  `build` jobs for my e2e apps to verify that they are valid, and while executing the tests I do not need the checks again/at all. Some history about that ... some time ago I had an typescript error in one of my e2e apps, but cypress ran without any problems - it just printed out the typescript error like cypress prints out any other error. I had expected that the CI job would have failed in another way. How are the errors that the `ForkTsCheckerWebpackPlugin` is detecting passed along? Are they? And would `enabled=false` help in my case to save memory?

**Comment by docwhite** [comment]
I ended up using a T3 Small (2 GB RAM) and I added to it 2 GB of swap, and now it likes it. I didn't need to play with any other option, but this obviously is not a solution. I made the CI run just 1 job at a time too, to avoid memory peaks.

**Comment by skydever** [comment]
@docwhite thx for the additional info! so you really run into hardware limits and by tweaking them up to about 4GB and with 1 job at a time it currently does not error?

the [config](https://github.com/TypeStrong/fork-ts-checker-webpack-plugin#typescript-options) of the `ForkTsCheckerWebpackPlugin` says that the default value is `2048 MB`, and I think that I do not run into hardware limits concerning that default value - it seems my error occurs because the default value is not high enough? 🤔 do you think it will just use less memory for the same job by reducing that value? if that would not be the case (and it seems like it is not), you will still error on a machine with only 2 GB of RAM!? But I just guess here ...

I also read at the [docs](https://github.com/TypeStrong/fork-ts-checker-webpack-plugin#options) of the `ForkTsCheckerWebpackPlugin` that it uses [cosmiconfig](https://github.com/davidtheclark/cosmiconfig), so we should be able to set the `memoryLimit` via additional configs (at `package.json` or additional config files) as a workaround for now. I will try that and let you know.

**Comment by docwhite** [comment]
Yes I capped the job concurrency to 1 to avoid that, I believe if 2 jobs can happen at once I would probably hit the error again. It can be mitigated by increasing your swap to something that's enough, but I believe it's better to tweak that config.

I didn't know it might be configurable with package.json / rcfile! It's a good route to explore and I would be happy to know your results.

Let me attach a table I did that shows the job times, you can see that you cannot have a tiny RAM and massive swap:

**Comment by skydever** [comment]
I added this to my `package.json` but it does not seem to work (I still get the error):

```
  "fork-ts-checker": {
    "typescript": {
      "memoryLimit": 8000
    }
  }
```

What did you do to get this output? Would help to verify what value is set :) I don't see this output 🤔 

```
Starting type checking service...
Using 1 worker with 2048MB memory limit
```

**Comment by docwhite** [comment]
> What did you do to get this output?

@skydever  I'm using [nx](https://nx.dev/) and I can run the tests on an Angular application by `nx run myapp-e2e:e2e --headless` or run all your e2e tests in your monorepo `nx run-many --target=e2e --all`.

**Comment by skydever** [comment]
@docwhite thx! I do it the same way, except with a small wrapper script around it that executes `nx run my-app:e2e:my-config` - but even when I execute `nx run my-app:e2e:my-config` directly I do not see that output (`my-config` is setting the `headless` option). I am not sure where this gets cut off - I am using a custom `reporter`, maybe that is why I do not see that output 🤔 ...

**Comment by skydever** [comment]
I just found out that the nx version I am using (12.5.7) is constructing the `ForkTsCheckerWebpackPlugin` like this:

```
new ForkTsCheckerWebpackPlugin({
  tsconfig: config.env.tsConfig,
  useTypescriptIncrementalApi: false,
})
```

so I have to set this at `package.json` I guess:

```
"fork-ts-checker": {
  "memoryLimit": 8000
}
```

I always looked at [preprocessor.ts](https://github.com/nrwl/nx/blob/master/packages/cypress/src/plugins/preprocessor.ts) of the current master, but that differs from 12.5.7 ... I will report my results again after trying ...

**Comment by skydever** [comment]
some more findings ... with nx 12.5.7 the version `3.1.1` of `ForkTsCheckerWebpackPlugin` is used, and this version does not use cosmiconfig :( ... I will upgrade to nx 12.5.8 and report afterwards ...

**Comment by docwhite** [comment]
@skydever Right! I thought `config.env.tsConfig` would refer to the `tsconfig.json` files! And that it couldn't be configurable from `package.json`! I don't know about the `cosmiconfig` magic.

**Comment by skydever** [comment]
hmm ... even after upgrading to nx `12.5.8` I get version `3.1.1` of `fork-ts-checker-webpack-plugin` (I use `npm`, not `yarn`) ... I think that's not correct!? 🤔 

my `package-lock.json` looks like this for `@nrwl/cypress`:

```
"@nrwl/cypress": {
      "version": "12.5.8",
      ...
      "requires": {
        "@cypress/webpack-preprocessor": "~4.1.2",
        "@nrwl/devkit": "12.5.8",
        "@nrwl/linter": "12.5.8",
        "@nrwl/workspace": "12.5.8",
        "fork-ts-checker-webpack-plugin": "^3.1.1", // hmm, where does this come from?
        "rxjs": "^6.5.4",
        "ts-loader": "5.4.5",
        "tsconfig-paths-webpack-plugin": "3.2.0",
        "tslib": "^2.0.0",
        "webpack-node-externals": "1.7.2",
        "yargs-parser": "20.0.0"
      },
```

... and the code of the `preprocessor.js` in my local `node_modules` seems outdated too (`node_modules\@nrwl\cypress\src\plugins\preprocessor.js`):

```
new ForkTsCheckerWebpackPlugin({
  tsconfig: config.env.tsConfig,
  useTypescriptIncrementalApi: false,
})
```

I do not use `@nrwl/node` or `@nrwl/web` that are using `ForkTsCheckerWebbackPlugin` too (if the dep. is controlled via one of those deps) ...

**Comment by skydever** [comment]
I think this *confusion* has something to do with the temp. solution for nx `13` like mentioned at #5892

something I tried anyway  ... I *hacked* my local `node_modules\@nrwl\cypress\src\plugins\preprocessor.js` to

```
new ForkTsCheckerWebpackPlugin({
  tsconfig: config.env.tsConfig,
  useTypescriptIncrementalApi: false,
  memoryLimit: 8000
})
```

to set a `memoryLimit`, but my tests still fail with the same error ...

**Comment by docwhite** [comment]
@skydever Make sure it's not another problem. Check the memory limit is propagating correctly, try adding a `console.log(this.memoryLimit)` after:

```javascript
        this.memoryLimit =
            options.memoryLimit || ForkTsCheckerWebpackPlugin.DEFAULT_MEMORY_LIMIT;
```

in `node_modules/fork-ts-checker-webpack-plugin/lib/index.ts` in the constructor method. You might end up discarding `memoryLimit` as the culprit maybe...

**Comment by skydever** [comment]
I have some great news 😄  ... nx version `12.6.0-beta.6` came out yesterday and I did the migration today. now version `6.2.1` of `fork-ts-checker-webpack-plugin` is used, and I was able to set the memory limit like I tried before (using [cosmoconfig](https://github.com/davidtheclark/cosmiconfig)) at `package.json`:

```json
  "fork-ts-checker": {
    "typescript": {
      "memoryLimit": 8000
    }
  }
```

I added a `console.log` at my local `\node_modules\fork-ts-checker-webpack-plugin\lib\ForkTsCheckerWebpackPlugin.js` to verify that the `memoryLimit` will be used, and I got the expected output - and now my tests passed, no out of memory error 🎉 

To sum up - `memoryLimit` should be added to the cypress config of nx to avoid the temp. workaround using [cosmoconfig](https://github.com/davidtheclark/cosmiconfig) that is possible with version `12.6.0-beta.6` of nx (but be aware, the option will be considered for each instance of `ForkTsCheckerWebpackPlugin` if you use this workaround) 👍

**Comment by docwhite** [comment]
Great! So they already anticipated our problem! I will wait until it hits production release. Thanks for sharing your investigation @skydever.

**Comment by barbados-clemens** [comment]
Hi there, have y'all tried updating to the latest version of cypress or something > 7, as the preprocessor plugin isn't needed. This will be removed in nx v14 as well. Let me know if it's still an issue. Thanks!
